Meadowlark-js/packages/meadowlark-authz-server/src/handler/CreateClient.ts

```diff
@@ -0,0 +1,104 @@
+// See the LICENSE and NOTICES files in the project root for more information.
+import crypto from 'node:crypto';
+import { v4 as uuidv4 } from 'uuid';
+import { authorizationHeader, LOCATION_HEADER_NAME, Logger } from '@edfi/meadowlark-core';
+import { CreateAuthorizationClientRequest } from '../message/CreateAuthorizationClientRequest';
+import { CreateAuthorizationClientResult } from '../message/CreateAuthorizationClientResult';
+import { CreateClientBody } from '../model/CreateClientBody';
+import { ensurePluginsLoaded, getAuthorizationStore } from '../plugin/AuthorizationPluginLoader';
+import { checkForAuthorizationErrors } from '../security/JwtValidator';
+import { AuthorizationRequest } from './AuthorizationRequest';
+import { AuthorizationResponse } from './AuthorizationResponse';
+import { BodyValidation, validateCreateClientBody } from '../validation/ValidateBody';
+import { CreateClientResponseBody } from '../model/CreateClientResponseBody';
+import { writeDebugStatusToLog, writeErrorToLog, writeRequestToLog } from '../Logger';
+
+const moduleName = 'handler.CreateClient';
+
+/**
+ * Handler for client creation
+ */
+export async function createClient(authorizationRequest: AuthorizationRequest): Promise<AuthorizationResponse> {
+  try {
+    writeRequestToLog(moduleName, authorizationRequest, 'createClient');
+    await ensurePluginsLoaded();
+
+    const errorResponse: AuthorizationResponse | undefined = checkForAuthorizationErrors(
+      authorizationHeader(authorizationRequest.headers),
+    );
+
+    if (errorResponse != null) {
+      writeDebugStatusToLog(moduleName, authorizationRequest, 'createClient', errorResponse.statusCode, errorResponse.body);
+      return errorResponse;
+    }
+
+    if (authorizationRequest.body == null) {
+      const message = 'Missing body';
+      writeDebugStatusToLog(moduleName, authorizationRequest, 'createClient', 400, message);
+      return { body: JSON.stringify({ message }), statusCode: 400 };
+    }
+
+    let parsedBody: any = {};
+    try {
+      parsedBody = JSON.parse(authorizationRequest.body);
+    } catch (error) {
+      const message = 'Malformed body';
+      writeDebugStatusToLog(moduleName, authorizationRequest, 'createClient', 400, message);
+      return { body: JSON.stringify({ message }), statusCode: 400 };
+    }
+
+    const validation: BodyValidation = validateCreateClientBody(parsedBody);
+    if (!validation.isValid) {
+      writeDebugStatusToLog(moduleName, authorizationRequest, 'createClient', 400, validation.failureMessage);
+      return {
+        body: validation.failureMessage,
+        statusCode: 400,
+      };
+    }
+
+    const createClientBody: CreateClientBody = parsedBody as CreateClientBody;
+
+    const clientId: string = uuidv4();
+    const clientSecretBytes: Buffer = crypto.randomBytes(32);
+    const clientSecretHashed: string = crypto.createHash('shake256').update(clientSecretBytes).digest('hex');
```

Comment:
If this were a user password, rather than a client secret, I would say we 100% [must have salt](https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/). But since this is already a randomly generated secret, I think we're OK as is. Still researching though.

Reply:
Not a security expert, but I don't think salt is necessary here as it's used to reduce effectiveness of pre-computed hashes of known password lists. No precomputable input here.

Reply:
[OAuth 2.0 Threat Model and Security Considerations](https://www.rfc-editor.org/rfc/rfc6819): "If the credential lacks a reasonable entropy level (because it is a user password) an additional salt will harden the storage to prevent offline dictionary attacks"

Already have a cryptographically strong random secret, meeting the bar for "reasonable entropy level".